Source/Diagnostics/ReducedDiags/RhoMaximum.cpp

```diff
@@ -0,0 +1,169 @@
+    const auto & mypc = WarpX::GetInstance().GetPartContainer();
+
+    // get number of species (int)
+    const auto nSpecies = mypc.nSpecies();
+
+    // A vector to store the indices of charges species in mypc
+    amrex::Vector<int> indices_charged_species;
+
+    // number of charged species
+    int n_charged_species = 0;
+
+    for (int i = 0; i < nSpecies; ++i)
+    {
+        // Only charged species are relevant for this diag
+        if (mypc.GetParticleContainer(i).getCharge() != amrex::Real(0.))
+        {
+            indices_charged_species.push_back(i);
+            n_charged_species += 1;
+            for (int lev = 0; lev < nLevel; ++lev)
+            {
+                // Initialize functors for the charge density of each charged species
+                m_rho_functors[lev].push_back(std::make_unique<RhoFunctor>(lev, crse_ratio, i));
+            }
+        }
+    }
+
+    // get species names (std::vector<std::string>)
+    const auto species_names = mypc.GetSpeciesNames();
+
+    // Min and max of total rho + max of |rho| for each species
+    const int noutputs_per_level = 2+n_charged_species;
+    m_data.resize(nLevel*noutputs_per_level, amrex::Real(0.0));
+
+    if (amrex::ParallelDescriptor::IOProcessor())
+    {
+        if ( m_IsNotRestart )
+        {
+            // open file
+            std::ofstream ofs{m_path + m_rd_name + "." + m_extension,
+                std::ofstream::out | std::ofstream::app};
+            // write header row
+            ofs << "#";
+            ofs << "[1]step()";
+            ofs << m_sep;
+            ofs << "[2]time(s)";
+            constexpr int shift_max_rho = 3;
+            constexpr int shift_min_rho = 4;
+            constexpr int shift_first_species = 5;
+            for (int lev = 0; lev < nLevel; ++lev)
+            {
+                ofs << m_sep;
+                ofs << "[" + std::to_string(shift_max_rho+noutputs_per_level*lev) + "]";
+                ofs << "max_rho_lev"+std::to_string(lev)+" (C/m^3)";
+                ofs << m_sep;
+                ofs << "[" + std::to_string(shift_min_rho+noutputs_per_level*lev) + "]";
+                ofs << "min_rho_lev"+std::to_string(lev)+" (C/m^3)";
+                for (int i = 0; i < n_charged_species; ++i)
+                {
+                    ofs << m_sep;
+                    ofs << "[" + std::to_string(shift_first_species+i+noutputs_per_level*lev) + "]";
+                    ofs << "max_" + species_names[indices_charged_species[i]]
+                                  + "_|rho|_lev"+std::to_string(lev)+" (C/m^3)";
+                }
+            }
+            ofs << std::endl;
+            // close file
+            ofs.close();
+        }
+    }
+
+}
+// end constructor
+
+// function that computes maximum charge density values
+void RhoMaximum::ComputeDiags (int step)
+{
+    // Judge if the diags should be done
+    if (!m_intervals.contains(step+1)) { return; }
+
+    // get WarpX class object
```

Comment:
I would say something like `get a reference to WarpX instance`

Reply:
Sure, I'll update this comment (also in the other reduced diags).